### DISCUSSION TITLE
Welcome to OpenConstructionSimulator Discussions!

### DISCUSSION THREAD

**Discussion by qoopen0815** [Field-Robotics-Japan/OpenConstructionSimulator#7] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by dataplayer12** [comment]
Hello @qoopen0815

Thank you for creating this project! I am trying to use OCS but I don't know how to control the objects in the default demo scene. I can control basic movements like moving forward using keyboard but I cannot get the backhoe to dig. Is there a 説明書/manual for using OCS or a brief explanation of basic game controls in unity?

**Reply by qoopen0815** [reply]
@dataplayer12 , thank you for reply.
Have you tried using a gamepad to control the game?
OCS doesn't officially support keyboard operation, so I'd recommend you to use a [gamepad](https://github.com/qoopen0815/OpenConstructionSimulator/tree/main/Assets/OCS/OcsVehicle/Docs/HowToPlay).

The settings related to heavy equipment operation are in [InputSystem](https://github.com/qoopen0815/OpenConstructionSimulator/tree/main/Assets/OCS/OcsVehicle/Scripts/InputSystem), so if you want to use the keyboard, please set your favorite keys here to deal with it.

**Reply by dataplayer12** [reply]
Hi @qoopen0815
Thanks for the detailed response. I could get the game working with a controller. I don't need to use the keyboard. I will try to automate inputs by simulating the controller itself.

Having tried the game with the controller, I found two other limitations:

1. How to get the excavator arm/joint angle data out from the simulator? I think I will have to modify the source, but where exactly in the source code is all vehicle data available? I plan to add some lines to save all the raw data to a text or csv file.

2. After playing around with the default scene, I found that the excavator does not dig into the ground! I try to dig, but the bucket passes right through the ground and nothing is picked up in the bucket. The ground also remains completely flat, exactly as it was before. Am I doing something wrong? Are some modifications required to get digging working? I am using the same version of Unity as specified in readme.

**Reply by qoopen0815** [reply]
Hi, @dataplayer12 
Thank you for trying it.
I'm glad to hear that you had success.

Now, before I answer your question, let me make one comment.
The Open Construction Simulator repository has been forked to an organization called FRJ, where development continues.
So the version you tried may have been a bit old.
I have merged the latest branch of FRJ into the develop branch of my repository, so please try that.

**Reply by qoopen0815** [reply]
The following is an answer to your question.

1. There is no direct way to get the joint angle. Each object in the arm has a parent-child relationship, so if you want to get the joint angle, please create your own script to handle it.

2. Sorry, this is a bug. Could you please wait a moment? I'll try to fix it as soon as possible.

**Reply by dataplayer12** [reply]
Hi @qoopen0815 
Thanks for the explanation. I have switched to using the `devel` branch. I am waiting for you to fix the digging bug. Please fix it as soon as you can. Do you have any advice for controlling each vehicle separately? When I control the vehicles with a game controller, all vehicles move simultaneously. 
Thank you for your continued efforts in making this project better!!

**Reply by ojanck** [reply]
> The following is an answer to your question.
> 
> 1. There is no direct way to get the joint angle. Each object in the arm has a parent-child relationship, so if you want to get the joint angle, please create your own script to handle it.
> 2. Sorry, this is a bug. Could you please wait a moment? I'll try to fix it as soon as possible.

Hi @qoopen0815, new to this thread and I wanted to add another question in mention of the bug. I tried using the OCSTerrain with an BackHoe asset that I got. Here's the thing though, when I try to put the ExcavateLine child obj in the tip of my rear bucket, the ground suddenly digs by itself without me having to move the bucket. Is there a step I'm missing or is there any of the script that I need to tweak?

You can see what happened in [here](https://drive.google.com/file/d/1je0jPFbbpg9IIPk8TZ2SNs441GFNcqs2/view?usp=sharing)

**Reply by qoopen0815** [reply]
Thanks for your feedback. I will check it out this weekend.

**Comment by M-gbf** [comment]
Hi!

I need a simulator for an automated excavator project, where i can put sensors, camera and other things.
Could you tell me if this simulator meets that? Does it communicate with ROS2? What are the possible programming languages?
Where can I download the full version?

Best regards.
